### ISSUE TITLE
`no-array-index-key` not triggered when map function is called on Objects rather than JSX

### ISSUE THREAD

**Issue by ismailcherri** [jsx-eslint/eslint-plugin-react#3569]
Hi!
I was wondering if it is possible to detect the following case:

```js
export default function Items() {
  const items = [
    { name: "item 1" },
    { name: "item 2" },
    { name: "item 3" },
    { name: "item 4" },
    { name: "item 5" },
    { name: "item 6" },
  ];

  const mappedItems = items.map((item, index) => {
    return { ...item, id: index };
  });

  return (
    <>
      {mappedItems.map((item) => {
        return <div key={item.id}>{item.name}</div>;
      })}
    </>
  );
}
```

Eslint will not complain about this even though the map index is used as key for the object.

**Comment by ljharb** [comment]
It's not possible for eslint to catch cases like this in general, because `mappedItems` could be created any number of ways.

I think this is the sort of thing that just has to be left up to code review.
